Fixes #24271 - Edit Host as non-admin is slower than admin

kgaikwad opened the pull request:
For non-admin users, it takes more time in processing
common_parameters partials as comparative to admin.
Removed '.can?' method calls for checking edit & destroy
permissions on parameters and pre-calculated these permissions.

kgaikwad committed:
Fixes #24271 - Edit Host as non-admin is slower than admin

For non-admin users, it takes more time in processing
common_parameters partials as comparative to admin.
Removed '.can?' method calls for checking edit & destroy
permissions on parameters and pre-calculated these permissions.

kgaikwad commented:
@ShimShtein, 
I have pushed first version of parameter changes where I cached the `user.permissions` in `.can?` method and modified the code bit as we discussed.
Please have a look.
If there is scope of improvement, please feel free to add comments :-)

ares commented on app/services/authorizer.rb:
```diff
@@ -173,4 +173,14 @@ def base_ids
+  end
+
+  def cache_key
+    { :"user_#{user.id}" => "#{user.last_login_on.to_i}-#{user.updated_at.to_i}" }
```
Does this mean that user needs to relogin after admin changes his/her permissions to drop the cache?

ares reviewed (dismissed):
IIUC the suggested change, I'm concerned with the fact we cache permission per session. That wouldn't be huge problem in case we use authorizer everywhere, but it's not the case. There some permission checks would use cache and some not. That would lead to very confusing behavior. Could the caching be added only to host params instead of modifying general authorizer code? Or is there other alternative? Or am I misreading the code? :-)

ShimShtein commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -70,9 +70,38 @@ def lookup_key_field(id, value, options)
+    return result if user.admin?
+    resource_class = parameters_by_type.klass
+    result.merge(
+      ids_to_edit: authorizer.cached_collection(resource_class, :edit_params).map(&:id),
```
I think we can do the same thing we do with `:view_params`:
``` ruby
{
ids_to_edit: parameters_by_type.authorized(:edit_params).pluck(:id)
ids_to_destroy: parameters_by_type.authorized(:destroy_params).pluck(:id)
}
```

ShimShtein commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -70,9 +70,38 @@ def lookup_key_field(id, value, options)
+  private
+
+  def find_allowed_param_ids_per_action(parameters_by_type)
+    result = { ids_to_view: parameters_by_type.authorized(:view_params).map(&:id) }
```
Instead of `#map` that works on returned collection of `Parameter` objects, it would be more efficient to use `#pluck`. It would generate a query that returns only the selected fields from the DB as an array.

ShimShtein commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -70,9 +70,38 @@ def lookup_key_field(id, value, options)
+  end
+
+  def action_allowed?(parameter, allowed_ids)
+    user = User.current
```
If we use `#authorized` scope, it wil automatically create `1=0` or `1=1` queries for these special cases. We shouldn't worry about it here.

ShimShtein reviewed (changes requested):
It's a good start, but if we use `#authorized` for every action, I think we won't need changes to `autorizer` at all. It will cause 3 queries to DB (which is acceptable as it's bound).

kgaikwad commented:
@ares, @ShimShtein,

Thank you for review.

> Does this mean that user needs to relogin after admin changes his/her permissions to drop the cache?


@ares, I agree with your point that there are places where it won't use cache result. so it may cause inconsistency.
Secondly, on changes into permissions of any role, I will need to touch all users object to get changed `updated_at` value. If I skip `touch` for user objects on this change, `user_id & updated_at` are not enough, so added `last_login_on` in cache_key. Now, realized that it will mislead
So I will remove this part and modify only for host_params i.e for `authorizer.can?(:create_params)`



> It's a good start, but if we use #authorized for every action, I think we won't need changes to authorizer at all. It will cause 3 queries to DB (which is acceptable as it's bound).

Yes, it will cause 3 queries to DB. So, every time we visit this page, it fires 3 queries.
Initially, I have used in a same way like `:view_params` but using `authorizer`  we can use existing cache logic to fetch result. So, later I modified it to `authorizer`. 
@ShimShtein, sorry that I am asking it again but please confirm do you want me to change it to ` #authorized`?

kgaikwad commented:
@ares, @ShimShtein,
Updated  PR with suggested changes.

@ShimShtein, I am not really sure whether we can merge this logic in one [query](https://github.com/theforeman/foreman/pull/5994/files#diff-d7466ddd9140741130fc23515bce126cR93) as method [find_collection](https://github.com/theforeman/foreman/blob/develop/app/services/authorizer.rb#L27) uses multiple scopes to get that result. I am checking on it. If any pointers to improve, please let me know.

ares commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -70,9 +70,31 @@ def lookup_key_field(id, value, options)
+
+  private
+
+  def find_allowed_param_ids_per_action(parameters_by_type)
```
I'm probably missins something and I'm reading this on phone but IIRC, authorizer exists exactly for this. It has separate caches for separate permissions. When it's instantiated, it can get preloaded set of objects. So I don't understand why we use it for create but bot for edit/destroy permissions here. If the collection of objects is different, we can have 3 authorizer instances.

ShimShtein replied:
@ares, IIRC where this issue started from, the problem was that the cache is not filtered at all - you were getting all parameters from the DB, so it was too expensive to build that cache.

kgaikwad replied:
@ShimShtein,

There are couple of changes suggested by @ares.
Basically, it includes instead of all parameters from the DB, it will limit to host's parameters.

ShimShtein commented:
I didn't test it yet, but from code perspective this looks good.

kgaikwad committed:
Refs #24271 - commit 2 with slower query change

kgaikwad commented:
@ShimShtein, @ares, 

I have added changes by passing collection of specific host_parameters while creation of authorizer object.
Please let me know these  changes are valid. Especially [lines](https://github.com/theforeman/foreman/pull/5994/files#diff-2a8c1ced7ef4dfe08f35154c5a3ef0f1R28) that I have added in `find_collection` method.

ShimShtein commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -70,9 +70,14 @@ def lookup_key_field(id, value, options)
-    resource_parameters = parameters_by_type.select { |p| parameter_ids_to_view.include?(p.id) }
+    @authorizer_for_params = Authorizer.new(User.current, :collection => parameters_by_type)
+    resource_parameters = @authorizer_for_params.find_collection(
+      parameters_by_type, :permission => :view_params)
```
I think that `parameters_by_type` should be replaced by parameter type here. `find_collection` needs a resource type as the first parameter, not a collection.

ShimShtein commented on app/services/authorizer.rb:
```diff
@@ -25,6 +25,9 @@ def can?(permission, subject = nil, cache = true)
   end
 
   def find_collection(resource_class, options = {})
+    return resource_class.where('1=0') if user.nil?
```
I think `find_collection` already knows how to short circuit these special cases in `build_filtered_scope_components`

kgaikwad replied:
I think, likewise in [authorizable.rb](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/authorizable.rb#L63), I need to move these two lines in helper file.

ares replied:
`find_collection` would only be called from `can?` method, which already has a [guard for this case](https://github.com/theforeman/foreman/blob/ef4b3b25f8874b7414616acac473472fc22d0d61/app/services/authorizer.rb#L13-L15)

edit: oh sorry I missed you're calling find_collection directly

ShimShtein replied:
@ares, i this case, should we leave the short circuit, or we can rely on `build_filtered_scope_components` to do the right job?

ares replied:
if Authorizer#find_collection will be used, I think I'd add it explicitly here, but after yesterday chat with @kgaikwad, maybe this won't be necessary

ShimShtein reviewed (changes requested):
This is a reasonable solution, I quite like it. It will need a couple of performance tests before we can merge it though.

ShimShtein commented:
@kgaikwad, what's the status here?

kgaikwad committed:
Refs #24271 - extracted user check outside of authorizer

kgaikwad commented:
@ShimShtein, @ares,
Updated PR with changes where user check extracted out in separate method before calling `find_collection` in helper file.

After discussion with @ares on IRC, he suggested alternative way which I think same as what I did using `.authorized` and calculating ids per action i.e edit, destroy, view.
I am going to test both solutions against parameter data. After that, it will be easy to get which solution will be best suited here.

kgaikwad commented:
@ares, @ShimShtein,
For these changes, I did some performance test with admin/non-admin user. Host edit action called twice so giving two time values for rendering partial. Here is a short summary:

**1. With current changes from develop branch.**
For admin user,
Rendered common_parameters/_parameters.html.erb (30.7ms)
Rendered common_parameters/_parameters.html.erb (28.2ms)
For non-admin user,
 Rendered common_parameters/_parameters.html.erb (32582.8ms)
 Rendered common_parameters/_parameters.html.erb (34845.5ms)

**2. With latest changes in this [pull-request](https://github.com/theforeman/foreman/pull/5994/commits/17fd8ff422d44b2f60d35ff72286c20a542a6149)**
For admin user,
Rendered common_parameters/_parameters.html.erb (32.0ms)
Rendered common_parameters/_parameters.html.erb (29.1ms)

For non-admin user, 
run 1st:
Rendered common_parameters/_parameters.html.erb (92.6ms)
Rendered common_parameters/_parameters.html.erb (43.2ms)
run 2nd:
Rendered common_parameters/_parameters.html.erb (48.8ms)
Rendered common_parameters/_parameters.html.erb (45.6ms)

**3. With changes using 3 authorizer objects in [first commit of PR](https://github.com/theforeman/foreman/pull/5994/commits/13983dd7e42f5dad506e2dcb4815d6e390595f65)**

For admin user, 
Rendered common_parameters/_parameters.html.erb (33.3ms)
Rendered common_parameters/_parameters.html.erb (34.8ms)

For non-admin user,
run 1st:
Rendered common_parameters/_parameters.html.erb (44.5ms)
Rendered common_parameters/_parameters.html.erb (39.1ms)
run 2nd:
Rendered common_parameters/_parameters.html.erb (43.8ms)
Rendered common_parameters/_parameters.html.erb (65.5ms)

Please let me know your suggestions.

Updated: From above results, result from point (3) i.e. initial commit of this PR using 3 authorizer objects gives better result than remaining two.

ares commented:
Great work @kgaikwad, I think both options of the fix are comparable resource-wise. I'd prefer the later appproach (current version of PR) as it's using Authorizer code and does not introduce custom methods.

How many parameters you have defined for these benchmarks? FTR we already have integration tests (parameter_permissions.rb) for this, so probably no need to cover anything else.

kgaikwad commented:
> Great work @kgaikwad, I think both options of the fix are comparable resource-wise. I'd prefer the later appproach (current version of PR) as it's using Authorizer code and does not introduce custom methods.
> 
> How many parameters you have defined for these benchmarks? FTR we already have integration tests (parameter_permissions.rb) for this, so probably no need to cover anything else.

I have used `28 host_parameters` for one host machine.

justahero commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -70,9 +70,24 @@ def lookup_key_field(id, value, options)
   end
+
+  def apply_only_user_filter(user, parameters_by_type)
+    return parameters_by_type.where('1=0') if user.nil?
```
I would recommend to return `parameters_by_type.none` instead. It is an `ActiveRecord::Relation` acting as a Null Object that is still chainable, but does not generate a database query.

kgaikwad commented:
@justahero,
Thank you for review. Added change using `.none` and fixed tests.
Tests was failing previously as current parameter objects with validation failure replaced by result parameter collection of Authorizer instead of comparing existing parameter objects with new list for view permission.

kgaikwad committed:
Refs #24271 - declared params_authorizer in partial than helper

kgaikwad commented:
@ShimShtein,

I have updated PR with changes we discussed about authorizer declaration and deprecation warning message.

ShimShtein commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -68,11 +68,30 @@ def lookup_key_field(id, value, options)
+  def authorized_resource_parameters(params_authorizer, parameters_by_type)
+    unless params_authorizer.is_a?(Authorizer)
+      Foreman::Deprecation.deprecation_warning(
+        '1.22',
```
We are working on 1.21, so it will be deprecated at 1.23.

ShimShtein commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -68,11 +68,30 @@ def lookup_key_field(id, value, options)
+        'resource and type arguments are deprecated.'\
+        'Please modify method calling as per new definition.'
+      )
+      return nil
```
Since we don't want to break existing flow for the users, we should try and "guess" the new parameters from the old ones:

``` ruby
    unless params_authorizer.is_a?(Authorizer)
      Foreman::Deprecation.deprecation_warning(
          '1.23',
          "definition changed for method 'authorized_resource_parameters'."\
          'New definition accepts Authorizer object & parameters collection as arguments.'\
          'resource and type arguments are deprecated.'\
          'Please modify method calling as per new definition.'
      )
      # Assign old parameter names and gues new parameters.
      resource = params_authorizer
      type = parameters_by_type
      parameters_by_type = resource.send(type)
      params_authorizer = Authorizer.new(User.current, :collection => parameters_by_type)
    end
```

ShimShtein commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -68,11 +68,30 @@ def lookup_key_field(id, value, options)
+
+  def find_parameters_to_view(params_authorizer, parameters_by_type, user = User.current)
+    return parameters_by_type.none if user.nil?
+    return parameters_by_type.where(nil) if user&.admin?
```
you don't have to use safe call here. `user` is never `nil`.

ShimShtein commented on app/views/common_parameters/_parameters.html.erb:
```diff
@@ -8,7 +8,11 @@
-      <% authorized_resource_parameters(f.object, type).each do |parameter| %>
+      <%
+        parameters_by_type = f.object.send(type)
+        @params_authorizer = Authorizer.new(User.current, :collection => parameters_by_type)
```
You can create a local variable, and not a class one here, and then pass it to inner template as local:

```suggestion
        params_authorizer = Authorizer.new(User.current, :collection => parameters_by_type)
```

ShimShtein commented on app/views/common_parameters/_parameters.html.erb:
```diff
@@ -8,7 +8,11 @@
+      %>
+      <% authorized_resource_parameters(@params_authorizer, parameters_by_type).each do |parameter| %>
         <%= f.fields_for type, parameter do |builder| %>
           <%= render "common_parameters/parameter", :f => builder %>
```
Pass the authorizer as a local variable:

```suggestion
          <%= render "common_parameters/parameter", :f => builder, :authorizer => params_authorizer %>
```

ShimShtein commented on app/views/common_parameters/_parameter.html.erb:
```diff
@@ -1,5 +1,5 @@
 <tr class="fields <%= 'has-error' if f.object.errors.any? %>" id="<%= dom_id(f.object) + '_row' %>">
-  <% authorized = f.object.new_record? ? authorizer.can?(:create_params) : authorizer.can?(:edit_params, f.object) %>
+  <% authorized = f.object.new_record? ? can_create_parameter?(f.object, @params_authorizer) : @params_authorizer.can?(:edit_params, f.object) %>
```
if we pass authorizer as local variable in `_parameters.html.erb`, we can use it here then:

```suggestion
  <% authorized = f.object.new_record? ? can_create_parameter?(f.object, authorizer) : authorizer.can?(:edit_params, f.object) %>
```

ShimShtein commented on app/views/common_parameters/_parameter.html.erb:
```diff
@@ -8,7 +8,7 @@
   <td>
     <%= f.check_box(:hidden_value, :class => 'set_hidden_value hide', :checked => f.object.hidden_value?) if authorized %>
-    <%= link_to_remove_fields(_('Remove'), f) if f.object.new_record? || authorizer.can?(:destroy_params, f.object) %>
+    <%= link_to_remove_fields(_('Remove'), f) if f.object.new_record? || @params_authorizer.can?(:destroy_params, f.object) %>
```
same goes here:

```suggestion
    <%= link_to_remove_fields(_('Remove'), f) if f.object.new_record? || authorizer.can?(:destroy_params, f.object) %>
```

ShimShtein reviewed (changes requested):
Now it looks pretty solid. A couple of small changes and it's done.

kgaikwad committed:
Refs #24271 - used param_authorizer as a local variable

kgaikwad commented:
@ShimShtein, thank you for review! 
Updated PR as per your suggestions.

ShimShtein reviewed (dismissed):
Looks good from my point of view.
@ares, anything to add here?

ares reviewed (commented):
I haven't tested but looks reasonable, @ShimShtein if you did the testing, please go ahead and merge.

tbrisker commented on app/helpers/common_parameters_helper.rb:
```diff
@@ -68,11 +68,34 @@ def lookup_key_field(id, value, options)
+    params_authorizer.find_collection(parameters_by_type.klass, :permission => :view_params)
+  end
+
+  def can_create_parameter?(parameter, params_authorizer)
```
we don't actually need the `parameter` parameter here

tbrisker reviewed (changes requested):
Just a tiny comment inline, other that that looks like quite a great improvement!

kgaikwad commented:
@tbrisker,
Good catch :+1: 
Made a suggested change to PR.

EDIT-
It is showing me that "kgaikwad dismissed ShimShtein’s stale review via afb9ace 7 minutes ago" 
is it due to force-pushed?

tbrisker commented:
@kgaikwad yes, we have the repo set up so new pushes dismiss old approvals to make sure the approval is still valid. Thanks for updating this!

tbrisker reviewed (approved):
Thanks @kgaikwad !
